Userland/Libraries/LibWeb/IndexedDB/Internal/Algorithms.cpp

```diff
@@ -0,0 +1,120 @@
+
+// https://w3c.github.io/IndexedDB/#open-a-database-connection
+WebIDL::ExceptionOr<JS::NonnullGCPtr<IDBDatabase>> open_a_database_connection(JS::Realm& realm, StorageAPI::StorageKey storage_key, String name, Optional<u64> maybe_version, JS::NonnullGCPtr<IDBRequest> request)
+{
+    // 1. Let queue be the connection queue for storageKey and name.
+    auto& queue = ConnectionQueueHandler::for_key_and_name(storage_key, name);
+
+    // 2. Add request to queue.
+    queue.append(request);
+
+    // 3. Wait until all previous requests in queue have been processed.
+    HTML::main_thread_event_loop().spin_until(JS::create_heap_function(realm.vm().heap(), [queue, request]() {
+        return queue.first() == request;
+    }));
+
+    // 4. Let db be the database named name in storageKey, or null otherwise.
+    auto maybe_db = Database::for_key_and_name(storage_key, name);
+    JS::GCPtr<Database> db;
+
+    // 5. If version is undefined, let version be 1 if db is null, or db’s version otherwise.
+    auto version = maybe_version.value_or(maybe_db.has_value() ? maybe_db.value()->version() : 1);
+
+    // 6. If db is null, let db be a new database with name name, version 0 (zero), and with no object stores.
+    // If this fails for any reason, return an appropriate error (e.g. a "QuotaExceededError" or "UnknownError" DOMException).
+    if (!maybe_db.has_value()) {
+        auto maybe_database = Database::create_for_key_and_name(realm, storage_key, name);
+
+        if (maybe_database.is_error()) {
+            return WebIDL::OperationError::create(realm, "Unable to create a new database"_string);
+        }
+
+        db = maybe_database.release_value();
+    }
+
+    // 7. If db’s version is greater than version, return a newly created "VersionError" DOMException and abort these steps.
+    if (db->version() > version) {
+        return WebIDL::VersionError::create(realm, "Database version is greater than the requested version"_string);
+    }
+
+    // 8. Let connection be a new connection to db.
+    auto connection = IDBDatabase::create(realm, *db);
+
+    // 9. Set connection’s version to version.
+    connection->set_version(version);
+
+    // 10. If db’s version is less than version, then:
+    if (db->version() < version) {
+        // 1. Let openConnections be the set of all connections, except connection, associated with db.
+        auto open_connections = db->associated_connections_except(connection);
+
+        // 2. For each entry of openConnections that does not have its close pending flag set to true,
+        //    queue a task to fire a version change event named versionchange at entry with db’s version and version.
+        for (auto& entry : open_connections) {
+            if (!entry->close_pending()) {
+                // FIXME:
+                /*
+                HTML::queue_a_task(HTML::Task::Source::DatabaseAccess, nullptr, nullptr, JS::create_heap_function(realm.vm().heap(), [&realm, entry, db, version]() {
```

Comment:
commented out code again in ` LibWeb: Implement most of open_a_database_connection ` :(

Reply:
the code should be added in ` LibWeb: Implement and use fire_a_version_change_event ` instead.